// html-validate/dist/meta/element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetaCopyableProperty = exports.TextContent = void 0;
var TextContent;
(function (TextContent) {
    /* forbid node to have text content, inter-element whitespace is ignored */
    TextContent["NONE"] = "none";
    /* node can have text but not required too */
    TextContent["DEFAULT"] = "default";
    /* node requires text-nodes to be present (direct or by descendant) */
    TextContent["REQUIRED"] = "required";
    /* node requires accessible text (hidden text is ignored, tries to get text from accessibility tree) */
    TextContent["ACCESSIBLE"] = "accessible";
})(TextContent = exports.TextContent || (exports.TextContent = {}));
/**
 * Properties listed here can be copied (loaded) onto another element using
 * [[HtmlElement.loadMeta]].
 */
exports.MetaCopyableProperty = [
    "metadata",
    "flow",
    "sectioning",
    "heading",
    "phrasing",
    "embedded",
    "interactive",
    "transparent",
    "form",
    "labelable",
    "requiredAttributes",
    "attributes",
    "permittedContent",
    "permittedDescendants",
    "permittedOrder",
    "requiredAncestors",
    "requiredContent",
];
